infrastructure/testdriver/bidi/emulation/set_geolocation_override.https.html

```diff
@@ -0,0 +1,71 @@
+<!DOCTYPE html>
+<meta charset="utf-8"/>
+<title>TestDriver bidi.emulation.set_geolocation_override method</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/resources/testdriver.js?feature=bidi"></script>
+<script src="/resources/testdriver-vendor.js"></script>
+
+<script>
+    promise_setup(async () => {
+        // Ensure permission is granted before proceeding.
+        await test_driver.bidi.permissions.set_permission({
+            descriptor: {name: "geolocation"},
+            state: "granted",
+        });
+
+        // Ensure any previously set geolocation emulations are cleared.
+        await test_driver.bidi.emulation.set_geolocation_override(
+            {coordinates: null});
+    });
+
+    /** Get the current geolocation or error */
+    function get_current_geolocation() {
+        return new Promise(
+            resolve => window.navigator.geolocation.getCurrentPosition(
+                position => resolve(position.coords.toJSON()),
+                error => resolve({code: error.code, message: error.message}),
+                // Fail fast if geolocation is not available.
+                {timeout: 500}
+            ))
+    }
+
+    /** Asserts that relevant coordinate properties match */
+    function assert_coordinates_match(actual, expected) {
+        for (const key in expected) {
+            assert_equals(actual[key], expected[key],
+                `"${key}" should match the expected value ${expected[key]}`);
+        }
+    }
+
+    const SOME_COORDINATES = {
+        latitude: 52.51,
+        longitude: 13.39,
+        accuracy: 0.5,
+        altitude: 34,
+        altitudeAccuracy: 0.75,
+        heading: 180,
+        speed: 2.77
+    };
+
+    promise_test(async (t) => {
+        // Get the initial geolocation (might be error).
+        const initial_coords = await get_current_geolocation();
+
+        // Set the geolocation override
+        await test_driver.bidi.emulation.set_geolocation_override({
+            coordinates: SOME_COORDINATES
+        });
+
+        // Get the geolocation after setting the override.
+        const coords = await get_current_geolocation();
+        // Assert the coordinates match the override.
+        assert_coordinates_match(coords, SOME_COORDINATES);
+
+        // Clear the geolocation override.
+        await test_driver.bidi.emulation.set_geolocation_override(
+            {coordinates: null});
+        // Assert coordinates are set to the original value.
+        assert_coordinates_match(await get_current_geolocation(), initial_coords);
```

Comment:
The comment up top says that `initial_coords` might be an error, but if so this assert will fail, right? If an initial error should fail a test, can you make it fail earlier? Or update the comment if it's actually not possible?

Reply:
Maybe the wording is not clear. The result of `get_current_geolocation` can be an object with coordinates, or `{code: error.code, message: error.message}`. Anyway, we expect that removing emulation puts the system under the test to the original state, so this assertion should still be valid. This approach has a drawback in case of the system under the test can access the real geolocation which is changing, or the error is dynamically changing. If this will be a case, in future we can change the approach to verify the coordinates just changed. But  I believe this approach is a bit more robust.

Reply:
I see, so it's really asserting that the original state is restored, whatever it was. No problem then.